DTensor fast path: port return_and_correct_aliasing and inplace/out checks

swolchok opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #167475
* #166808
* #166372
* #167051

This seems to generate a several-microsecond performance improvement in the detach benchmark I've been using.

cc @H-Huang @awgu @wanchaol @fegin @fduwjj @wz337 @wconstab @d4l3k @pragupta @msaroufim @dcci

swolchok committed:
WIP: DTensor fast path: port return_and_correct_aliasing and inplace/out checks

This seems to generate a several-microsecond performance improvement in the detach benchmark I've been using.

[ghstack-poisoned]

swolchok committed:
fix apparent memory safety issue in #166808 on "WIP: DTensor fast path: port return_and_correct_aliasing and inplace/out checks"


This seems to generate a several-microsecond performance improvement in the detach benchmark I've been using.

cc H-Huang awgu wanchaol fegin fduwjj wz337 wconstab d4l3k pragupta msaroufim dcci

[ghstack-poisoned]

swolchok committed:
fix two test failures on "WIP: DTensor fast path: port return_and_correct_aliasing and inplace/out checks"


This seems to generate a several-microsecond performance improvement in the detach benchmark I've been using.

cc H-Huang awgu wanchaol fegin fduwjj wz337 wconstab d4l3k pragupta msaroufim dcci

[ghstack-poisoned]

swolchok committed:
fix bugs in #166808: failure to replace DTensors in OptionalTensorList arguments for local dispatch, and failure to return a list (was pushing multiple retvals onto stack) for list returning ops on "WIP: DTensor fast path: port return_and_correct_aliasing and inplace/out checks"


This seems to generate a several-microsecond performance improvement in the detach benchmark I've been using.

cc H-Huang awgu wanchaol fegin fduwjj wz337 wconstab d4l3k pragupta msaroufim dcci

[ghstack-poisoned]

swolchok committed:
fix a py::handle that should've been py::object on "WIP: DTensor fast path: port return_and_correct_aliasing and inplace/out checks"


This seems to generate a several-microsecond performance improvement in the detach benchmark I've been using.

cc H-Huang awgu wanchaol fegin fduwjj wz337 wconstab d4l3k pragupta msaroufim dcci

[ghstack-poisoned]

swolchok committed:
fix obvious brokenness/segfault on "WIP: DTensor fast path: port return_and_correct_aliasing and inplace/out checks"


This seems to generate a several-microsecond performance improvement in the detach benchmark I've been using.

cc H-Huang awgu wanchaol fegin fduwjj wz337 wconstab d4l3k pragupta msaroufim dcci

[ghstack-poisoned]

swolchok committed:
fix checking for default overloads on "WIP: DTensor fast path: port return_and_correct_aliasing and inplace/out checks"


This seems to generate a several-microsecond performance improvement in the detach benchmark I've been using.

cc H-Huang awgu wanchaol fegin fduwjj wz337 wconstab d4l3k pragupta msaroufim dcci

[ghstack-poisoned]

swolchok committed:
rebase to main to pick up #167580 on "DTensor fast path: port return_and_correct_aliasing and inplace/out checks"


This seems to generate a several-microsecond performance improvement in the detach benchmark I've been using.

cc H-Huang awgu wanchaol fegin fduwjj wz337 wconstab d4l3k pragupta msaroufim dcci

[ghstack-poisoned]

ezyang commented on torch/distributed/tensor/_op_schema.py:
```diff
@@ -478,10 +464,16 @@ def get_mesh_from_args(self, validate: bool = True) -> DeviceMesh:
+        # simple analysis of function schema to determine
+        # if this is an inplace variant, it might not
+        # be entirely correct, but it's good enough for now.
+        return self.op._schema.name[-1] == "_"
```
What happened here?

ezyang commented on torch/csrc/autograd/python_variable.cpp:
```diff
@@ -1539,7 +1572,96 @@ py::object dispatchDTensorOp(
+      }
+    }
+    stack->clear();
+    return wrapped_result;
```
This is probably possible to do better now that it's in C++, but FUTURE WORK.

swolchok committed:
rebase to viable/strict, fix some tail build issues and add a couple comments on "DTensor fast path: port return_and_correct_aliasing and inplace/out checks"


This seems to generate a several-microsecond performance improvement in the detach benchmark I've been using.

cc H-Huang awgu wanchaol fegin fduwjj wz337 wconstab d4l3k pragupta msaroufim dcci

[ghstack-poisoned]

swolchok committed:
fix mac tests at bottom of stack with USE_DISTRIBUTED gating in is_dtensor on "DTensor fast path: port return_and_correct_aliasing and inplace/out checks"


This seems to generate a several-microsecond performance improvement in the detach benchmark I've been using.

cc H-Huang awgu wanchaol fegin fduwjj wz337 wconstab d4l3k pragupta msaroufim dcci

[ghstack-poisoned]

swolchok committed:
fix mac tests at bottom of stack, second try on "DTensor fast path: port return_and_correct_aliasing and inplace/out checks"


This seems to generate a several-microsecond performance improvement in the detach benchmark I've been using.

cc H-Huang awgu wanchaol fegin fduwjj wz337 wconstab d4l3k pragupta msaroufim dcci

[ghstack-poisoned]

swolchok commented:
@pytorchbot merge -i
